### ISSUE TITLE
Event Handling

### ISSUE THREAD

**Issue by jalcine** [woboq/qmetaobject-rs#37]
Hey there! Firstly, thanks for this project. I'm using to build a [desktop client for social networking][1]. One thing that's critical to have it work well is event handling. There's a bit of work that one can do with signals using `qmetaobject` but I want to push the envelope by seeing if we can extend it to support things like `QEvent`. A more clinically description could be [found here][2]:

>  Does Qt let one plug pollable stuff into its main loop?  You may be able to steal some techniques from here:
>
> Motivation: https://coaxion.net/blog/2019/02/mpsc-channel-api-for-painless-usage-of-threads-with-gtk-in-rust/
>
> Implementation: https://github.com/gtk-rs/glib/blob/master/src/main_context_channel.rs
>
> TL;DR - have a channel be a bounded deque, with hooks into the main loop so the toolkit can ask it, "do you have events for me?".  That implementation exposes a GSource for a GMainLoop; not sure what Qt uses.

Thanks! I'll be tinkering on this in my projects and use it to potentially submit PRs.

[1]: https://github.com/blackaf/activitydesk
[2]: https://mstdn.mx/@federicomena/102210092521941421

**Comment by ogoffart** [comment]
I'm not sure I understand exactly what is needed here.

What there is so far is is `qmetaobject::queued_callback`

With that, you can do stuff such as:

```
let callback = queued_callback(move |x : String| {
       // this is run in the Qt main thread
      some_object.set_text(x)
});
    
std::thread::spawn(move || {
      thread::sleep(time::Duration::from_secs(10));
      callback("finished".into());
});
```

Not sure if that's what you are thinking of or if it can be improved.

**Comment by jalcine** [comment]
Heh, just as I posted this PR, I took a closer look at the source and noticed that method. The example that I have in mind is something like the following path:

  * App opens up a [channel](https://doc.rust-lang.org/std/sync/mpsc/)
  * App takes a clone of a [Sender](https://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html) and does some work in a background thread and builds up a list.
  * App now wants to use information from the background thread to populate a ListModel.



From what you have there, I was thinking of firing off a timer on the Qt main thread every so far to ‘flush’ a list populated from the background thread but if you know of a better, more reactive way; I’d love to hear it.

(Originally published at: https://v2.jacky.wtf/post/96107180-eb05-464b-a65e-26a58b30bcd6)

**Comment by jalcine** [comment]
This looks close. It seems like I needed to get into using `QPointer` more! That allowed me to move my use of the delegate across threads. The only issue now that I’m running into is that `QPointer` doesn’t allow for mutations. Without this, I can’t update my list model (thus providing people with an empty feed, heh).

(Originally published at: https://v2.jacky.wtf/post/02446ddd-1944-434b-a874-43adf3afe28a)

**Comment by ogoffart** [comment]
You can use `QPointer::as_pinned().borrow_mut()`  to get a mutable reference.

I understand this API is not the most obvious and the documentation is also lacking.

**Comment by jalcine** [comment]
No worries! I’m opting to generate most of it locally for now until that other ticket with generation remotely is resolved.

(Originally published at: https://v2.jacky.wtf/post/c05c3b35-bcd9-4ada-9a1a-e7341276ed57)
